File: 236.lowest-common-ancestor-of-a-binary-tree.py
#
# @lc app=leetcode id=236 lang=python3
#
# [236] Lowest Common Ancestor of a Binary Tree
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def lowestCommonAncestor(
        self, root: "TreeNode", p: "TreeNode", q: "TreeNode"
    ) -> "TreeNode":
        # cannot find LCA
        if not root:
            return None
        # some possibilities the answer is root itself
        if root == p or root == q:
            return root
        # the answer is from left child?
        left = self.lowestCommonAncestor(root.left, p, q)
        # the answer is from right child?
        right = self.lowestCommonAncestor(root.right, p, q)
        # if it's not from left child, the answer must at right child
        if left == None:
            return right
        # it's not from right child, the answer must at left child
        if right == None:
            return left
        # it's both left and right have LCA answer,
        # it means p and q locate at two different sides of node, the answer is root
        else:
            return root


# @lc code=end

File: 236.lowest-common-ancestor-of-a-binary-tree.go
/*
 * @lc app=leetcode id=236 lang=golang
 *
 * [236] Lowest Common Ancestor of a Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

package main

/*
 0. if node is nil, return nil
 1. p is root -> return p
 2. q is root -> return q
 3. neither p nor q is root
    a. two of them are in left tree -> if got nil, means two nodes in right tree, return right answer
    b. two of them are in right tree -> if got nil, means two nodes in left tree, return left answer
    c. if not the above cases (both answer return a node), return root
*/
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	if root.Val == p.Val {
		return p
	}
	if root.Val == q.Val {
		return q
	}
	leftAns := lowestCommonAncestor(root.Left, p, q)
	rightAns := lowestCommonAncestor(root.Right, p, q)
	if leftAns == nil {
		return rightAns
	}
	if rightAns == nil {
		return leftAns
	}
	return root
}

// @lc code=end
